e2e/tests/upgrades/upgrade_test.go

```diff
@@ -386,6 +391,147 @@ func (s *UpgradeTestSuite) TestV5ToV6ChainUpgrade() {
 	})
 }
 
+func (s *UpgradeTestSuite) TestV6ToV7ChainUpgrade() {
+	t := s.T()
+	testCfg := testconfig.FromEnv()
+
+	ctx := context.Background()
+	relayer, channelA := s.SetupChainsRelayerAndChannel(ctx)
+	chainA, chainB := s.GetChains()
+
+	createClientOptions := ibc.CreateClientOptions{
+		TrustingPeriod: time.Duration(time.Hour * 24 * 7).String(),
+	}
+
+	// create second tendermint client
+	s.SetupClients(ctx, relayer, createClientOptions)
+
+	status, err := s.Status(ctx, chainA, ibctesting.FirstClientID)
+	s.Require().NoError(err)
+	s.Require().Equal("Active", status)
+
+	status, err = s.Status(ctx, chainA, ibctesting.SecondClientID)
+	s.Require().NoError(err)
+	s.Require().Equal("Active", status)
+
+	var (
+		chainADenom         = chainA.Config().Denom
+		chainBIBCToken      = testsuite.GetIBCToken(chainADenom, channelA.Counterparty.PortID, channelA.Counterparty.ChannelID) // IBC token sent to chainB
+		soloMachineClientID = "06-solomachine-2"
+	)
+
+	chainAWallet := s.CreateUserOnChainA(ctx, testvalues.StartingTokenAmount)
+	chainAAddress := chainAWallet.Bech32Address(chainA.Config().Bech32Prefix)
+
+	chainBWallet := s.CreateUserOnChainB(ctx, testvalues.StartingTokenAmount)
+	chainBAddress := chainBWallet.Bech32Address(chainB.Config().Bech32Prefix)
+
+	// create solo machine client
+	solo := ibctesting.NewSolomachine(t, testsuite.Codec(), "solomachine", "testing", 1)
+
+	legacyConsensusState := &v7.ConsensusState{
+		PublicKey:   solo.ConsensusState().PublicKey,
+		Diversifier: solo.ConsensusState().Diversifier,
+		Timestamp:   solo.ConsensusState().Timestamp,
+	}
+
+	legacyClientState := &v7.ClientState{
+		Sequence:                 solo.ClientState().Sequence,
+		IsFrozen:                 solo.ClientState().IsFrozen,
+		ConsensusState:           legacyConsensusState,
+		AllowUpdateAfterProposal: true,
+	}
+
+	msgCreateSoloMachineClient, err := clienttypes.NewMsgCreateClient(legacyClientState, legacyConsensusState, chainAAddress)
+	s.Require().NoError(err)
+
+	resp, err := s.BroadcastMessages(
+		ctx,
+		chainA,
+		chainAWallet,
+		msgCreateSoloMachineClient,
+	)
+
+	s.AssertValidTxResponse(resp)
+	s.Require().NoError(err)
+
+	status, err = s.Status(ctx, chainA, soloMachineClientID)
+	s.Require().NoError(err)
+	s.Require().Equal("Active", status)
+
+	res, err := s.ClientState(ctx, chainA, soloMachineClientID)
+	s.Require().NoError(err)
+	s.Require().Equal(res.ClientState.TypeUrl, "/ibc.lightclients.solomachine.v2.ClientState")
```

Comment:
I think these should be swapped? Usually the expected value comes first. We could potentially use v7.ClientState{}.TypeURL()? (not sure if that compiles? or how to access the type url given a concrete type, might need to be casted to a proto.Message). I'm okay using a string but then we should make a const at the top of this file for v2 and v3 and state that the upgrade from v6 to v7 migrates the solo machine from v2 to v3 (assume the test reader has no context on the migrations performed)